java-squid/src/main/java/org/sonar/java/resolve/Types.java

```diff
@@ -84,4 +90,107 @@ public boolean isSubtype(JavaType t, JavaType s) {
     return result;
   }
 
+  /**
+   * Compute the "Least Upper Bound" ("lub", jls8 §4.10.4) of a list of type. The "lub" is a shared supertype that is more specific than any
+   * other shared supertype (that is, no other shared supertype is a subtype of the least upper bound)
+   *
+   * FIXME Parameterized types are currently ignored, as the method is used only to handle Union Types Trees, themselves being used only
+   * in catch trees. Note that Exceptions (any subclass of Throwable) cannot be generic (jls8 §8.1.2, §11.1.1: "compile-time error if a generic
+   * class is a direct or indirect subclass of Throwable")
+   *
+   * @param types
+   * @return the least upper bound of the types
+   */
+  public Type leastUpperBound(Type... types) {
+    Preconditions.checkArgument(types.length > 1);
+
+    List<List<Type>> supertypes = supertypes(types);
+    if (containsParameterizedTypes(supertypes)) {
+      // Parameterized types are ignored, erased supertypes should be calculated
+      return Symbols.unknownType;
```

Comment:
This is actually deadcode regarding where it is used, so I would not return unknown but actually fail in this case if you want to cover it.